proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,634 @@
+for resolving this: `PackageImports`. An import with a package specifier should look
+in the *non*-merged virtual module from that package.
+
+However, there is also a scoping
+problem. If we want the dotted-name import ``import Data.Set`` to work
+as a qualified-name import, then there must be a module ``Data`` *in scope*
+at that point. But we probably don’t want ``Data`` to be unconditionally
+in scope in the whole module! (Or do we?)
+
+.. [#] This is the case in Scala, for example.
+
+One possibility would be to have a special scope in the file module
+header where the “external modules” are in scope. I don’t know whether
+this would be nicer overall.
+
+Do we need different syntax for "local" imports?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Consider:
+
+::
+
+   module A where
+       import qualified B
+       module C where
+          import B.C
+
+From a user perspective, the operations carried out in the two import
+statements look very similar. However, there are differences because of the
+way we're handling dotted names and file module imports. We *could* signal this
+by having a different syntax for the new kind of imports, e.g. ``open`` (to mimic ML).
+I don't *think* this is necessary, but I might be wrong, and it might be clearer
+anyway.
+
+Do we need additional syntax for re-exporting all bindings from a module?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This proposal suggests using the export specifier ``module M(..)`` to re-export all the
+names defined in ``M``. This is useful when you want to effectively just rearrange the
+module structure without changing what is in a particular module.
+
+However, by analogy with the export specifier ``C(..)``, this should also export ``M``.
+I'm not sure if this is what we want, and if we don't, then perhaps we should have a
+different syntax.
+
+Datalog implementation
+----------------------
+
+The text of this proposal is mostly expository. I have written a model
+of the rules for the old and the new system in Souffle
+[souffle]_, a Datalog variant, which is a
+nice format for writing executable logical rules. This can be found at
+[hsmods-logic]_.
+
+So far, I have only implemented the simplest versions of the systems, in
+particular ignoring qualified imports and import/export specifiers. I
+will add qualified imports given more time, but I don’t plan to add
+import/export specifiers, since they aren’t very interesting and just
+act as filters on the imported/exported bindings.
+
+Since it is a logic programming language, the Datalog rules could easily
+be turned into inference rules to give a formal specification of the
+scoping relations of Haskell.
+
+The Datalog implementation also provides *an* implementation strategy: we
+can directly copy the evaluation model of Datalog. However, this is
+unlikely to be what we would do in a real compiler, and a procedural
+version is certainly possible.
```

Comment:
Datalog nudges us towards principled incremental compilation. Probably not a short term benefit, but differently a long term one. CC @kmicklas @ekmett.

Reply:
I didn't call it out in the proposal but this is true. Although if you wanted to do incremental computation, you'd be modifying the input relations (e.g. `modDefines` & co), which isn't standard. Fortunately, I wrote [a paper](https://link.springer.com/chapter/10.1007/978-3-030-17184-1_19) on how to do that :)

Reply:
Oh! What a coincidence :)